app/src/test/java/com/pmatuki/wowpapers/usecases/GetDownloadApplyWallpaperTest.kt

```diff
@@ -0,0 +1,79 @@
+package com.pmatuki.wowpapers.usecases
+
+import com.pmatuki.wowpapers.data.WallpapersRepositoryImpl
+import com.pmatuki.wowpapers.fake.WallpaperApplierFake
+import com.pmatuki.wowpapers.fake.WallpaperDownloaderFake
+import com.pmatuki.wowpapers.fake.WallpaperListSourceFake
+import com.pmatuki.wowpapers.usecases.apply.ApplyWallpaper
+import com.pmatuki.wowpapers.usecases.apply.WallpaperApplyResult
+import com.pmatuki.wowpapers.usecases.apply.WallpaperItemHolder
+import com.pmatuki.wowpapers.usecases.download.DownloadResult
+import com.pmatuki.wowpapers.usecases.download.DownloadWallpaper
+import com.pmatuki.wowpapers.usecases.get.GetWallpapers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Assert
+import org.junit.Test
+
+@ExperimentalCoroutinesApi
+class GetDownloadApplyWallpaperTest {
+
+    private val wallpaperListSourceFake = WallpaperListSourceFake(GetWallpapersTest.WALLPAPER_COUNT)
+
+    private val getWallpapers = GetWallpapers(WallpapersRepositoryImpl(wallpaperListSourceFake))
+
+    private val downloadWallpaper = DownloadWallpaper(WallpaperDownloaderFake())
+
+    private val applyWallpaper = ApplyWallpaper<ByteArray>(WallpaperApplierFake())
+
+    @Test
+    fun `test list, download second item and apply successfully`()  = runBlockingTest {
+        // List
+        val wallpaperList = getWallpapers()
+        Assert.assertTrue(wallpaperList.count() == GetWallpapersTest.WALLPAPER_COUNT)
+
+        // Download
+        val downloadResult = downloadWallpaper(wallpaperList[2].url)
+        Assert.assertTrue(downloadResult is DownloadResult.Success<*>)
+        val downloadSuccess = (downloadResult as DownloadResult.Success<*>)
+        Assert.assertNotNull(downloadSuccess.item)
+
+        // Apply
+        val applyResult = applyWallpaper(downloadSuccess.item as WallpaperItemHolder<ByteArray>)
+        Assert.assertTrue(applyResult is WallpaperApplyResult.Success)
+    }
+
+    @Test
+    fun `test list, error on download fifth item`()  = runBlockingTest {
+        // List
+        val wallpaperList = getWallpapers()
+        Assert.assertTrue(wallpaperList.count() == GetWallpapersTest.WALLPAPER_COUNT)
+
+        // Download with error
+        val downloadWallpaperWithError = DownloadWallpaper(WallpaperDownloaderFake().apply { doError() })
```

Comment:
I personally think this is a complex approach for the desired output. In my opinion, I prefer something like
```
val downloaderFake = WallpaperDownloaderFake().apply { returnError() }
val downloadWallpaperWithError = DownloadWallpaper(downloaderFake)
```

Or even, for more flexibility:
```
val downloaderFake = WallpaperDownloaderFake().apply { state = State.Error }
val downloadWallpaperWithError = DownloadWallpaper(downloaderFake)
```

I really like test codes to be easier to read, so I prefer to break it in small parts in order to achieve better readability.

Reply:
The last approach is much easier to understand indeed! Thanks!